### DISCUSSION TITLE
How to render list of bytes into a QR-code

### DISCUSSION THREAD

**Discussion by justme26** [chillerlan/php-qrcode#143] [Q&A]
Hi,

How do I render a list of bytes into a QR-code? 

This is my code:
```

<?php
use chillerlan\QRCode\{QRCode, QROptions};
use chillerlan\QRCode\Data\QRMatrix;
use chillerlan\QRCode\Common\EccLevel;

include_once('vendor/autoload.php');

$byte_array = array(65,66,67,177,178,68);

$options = new QROptions([
    'version'      => 40  
]);


echo '<img src="'.(new QRCode($options))->render($byte_array).'" alt="QR Code" />';
```

But chillerlan\QRCode\QRCode::render() only accepts strings so I am getting a Fatal Error. 

How can I do this? 

I want to save as many bytes as possible into one QR-code.

**Comment by codemasher** [comment]
Hey, I have recently added a method to the QR Code reader result class that does exactly this:

https://github.com/chillerlan/php-qrcode/blob/739579552d54d1b24a75415af805faba37b975fd/src/Decoder/DecoderResult.php#L90-L96


You can then invoke an output interface with the matrix like so:
```php
$out = new QRMarkupSVG($options, $matrix);
$out->dump();
```

**Reply by justme26** [reply]
Thanks for your reply. I thought I should use a encoder in this case?

**Reply by codemasher** [reply]
That is right, the bit in the decoder result just encodes the raw bytes of the reading back into a matrix object so that it can be consumed by an output module ([or debugger](https://github.com/chillerlan/php-qrcode/blob/main/tests/QRCodeReaderTestAbstract.php#L84-L87)) again.

**Reply by codemasher** [reply]
After some consideration i've [added `QRData::setBitBuffer()`](https://github.com/chillerlan/php-qrcode/commit/9b36cfb4a5f07c685557be9cf0a872d93c338c17) as a convenience method, so that you can throw the `QRdata` object into the mask pattern evaluation etc. Please note that there is no version auto detection however. The code would look something like this then:

```php
// fill a BitBuffer instance with the raw bytes
$bitBuffer   = new BitBuffer($rawBytes);
// set options /w version
$options     = new QROptions(['version' => 3]);
// invoke QRData and set the BitBuffer instance
$qrData      = (new QRData($options))->setBitBuffer($bitBuffer);
// evaluate mask pattern
$maskPattern = MaskPattern::getBestPattern($qrData);
// get matrix instance
$matrix      = $qrData ->writeMatrix($maskPattern);
// invoke output
$output      = new QRMarkupSVG($options, $matrix);
echo $output->dump();
```
Profit!

**Reply by justme26** [reply]
I tried this:

> use chillerlan\QRCode\{QRCode, QROptions};
> use chillerlan\QRCode\Data\QRMatrix;
> use chillerlan\QRCode\Common;
> use chillerlan\QRCode\Common\EccLevel;
> use chillerlan\QRCode\Decoder;
> include_once('vendor/autoload.php');
> 
> $rawBytes=array(65,66,67,177,178,68);
> 
> // fill a BitBuffer instance with the raw bytes
> $bitBuffer   = new chillerlan\QRCode\Common\BitBuffer($rawBytes);
> // set options /w version
> $options     = new QROptions(['version' => 3]);
> // invoke QRData and set the BitBuffer instance
> $qrData      = (new chillerlan\QRCode\Data\QRData($options))->setBitBuffer($bitBuffer);
> // evaluate mask pattern
> $maskPattern = chillerlan\QRCode\Common\MaskPattern::getBestPattern($qrData);
> // get matrix instance
> $matrix      = $qrData ->writeMatrix($maskPattern);
> // invoke output
> $output      = new chillerlan\QRCode\Output\QRMarkupSVG($options, $matrix);
> echo $output->dump();
> 

But i am getting a list of errors like this: 
Warning: Undefined array key 6 in /var/www/html/barcode/vendor/chillerlan/php-qrcode/src/Common/ReedSolomonEncoder.php on line 52

...and the outcoming QR-code is not readable :(

**Reply by codemasher** [reply]
That may happen when the given version does not match the given byte array. I've [tested it](https://github.com/chillerlan/php-qrcode/blob/9b36cfb4a5f07c685557be9cf0a872d93c338c17/tests/Data/QRDataTest.php#L29-L61) with several values and it seems that the "undefined array key" error happens when the version number is too high.

**Reply by codemasher** [reply]
An update: the version needs to match the [length bits range](https://github.com/chillerlan/php-qrcode/blob/main/src/Common/Mode.php#L42-L52) for the data mode the data has been encoded with, additionally the array needs to contain enough [pad bits](https://github.com/chillerlan/php-qrcode/blob/b8f287183775b51dff1f9f615b956d3c251dc084/src/Data/QRData.php#L208-L229).

**Reply by justme26** [reply]
When I use your code it also works on my machine for version 3. But when I use version 40 the QR-code can not be read by my QR-code reader (Binary Eye - phone app). Also, I have to ask how the rawData bytes is being translated to the link :)

**Reply by justme26** [reply]
Maybe I need to set "length bits range" to byte (8 bit)

**Reply by justme26** [reply]
... and then length of rawData should be 2953?

**Reply by codemasher** [reply]
> [...] works on my machine for version 3. But when I use version 40 [...]

it won't work on any other machine too because of the non-matching length bits. if it works for version 3, the maximum version would be 9 as this is the first breakpoint.

**Reply by justme26** [reply]
Sorry, I am a bit confused.... Is it possible to store 2953 bytes (or 2953*8 bits) in a single QR-code when using version 40?

**Reply by codemasher** [reply]
As per specification it is. https://www.qrcode.com/en/about/version.html
2952 bytes in binary mode, as the terminator consumes 4 bits at least and the data mode identifier another 4.

```php
	(new QRCode)->render(str_repeat('#', 2952)); // 2953 will yield an error
```

**Reply by justme26** [reply]
Okay, thanks :) Can you tell me how to select binary mode?

**Reply by codemasher** [reply]
It's as simple as that (in dev-main/v5):

```php
$qrcode = new QRCode($options);
$qrcode->addByteSegment($data);

$qrcode->render();
```

**Reply by codemasher** [reply]
The example before used the internal auto detection which will automatically switch to binary mode if no matching characters for numeric, alphanum or kanji are found.
